File: android/core/src/main/java/com/americangroupllc/drift/core/obs/AnalyticsService.kt
package com.americangroupllc.drift.core.obs

/** Capture surface that triggered an event. Mirrors Swift's `Surface` enum. */
enum class Surface(val raw: String) {
    APP("app"),
    NOTIFICATION_EXTENSION("notification_extension"),
    WATCH("watch"),
    COMPLICATION("complication"),
    TILE("tile"),
}

/** Drift's canonical event taxonomy. Same names fire from Swift + Kotlin. */
sealed class AnalyticsEvent(val name: String, val properties: Map<String, String>) {
    object OnboardingCompleted : AnalyticsEvent("onboarding_completed", emptyMap())

    class WaveSent(layer: String, surface: Surface) :
        AnalyticsEvent("wave_sent", mapOf("layer" to layer, "surface" to surface.raw))

    class WaveMatched(layer: String, timeToMatchSeconds: Long) :
        AnalyticsEvent("wave_matched", mapOf("layer" to layer, "time_to_match_seconds" to timeToMatchSeconds.toString()))

    class ChatScreenOpen(conversationId: String, tone: String) :
        AnalyticsEvent("chat_screen_open", mapOf("conversation_id" to conversationId, "tone" to tone))

    class ReplySuggestionUsed(tone: String, kind: String) :
        AnalyticsEvent("reply_suggestion_used", mapOf("tone" to tone, "kind" to kind))

    class ReplySuggestionDismissed(tone: String) :
        AnalyticsEvent("reply_suggestion_dismissed", mapOf("tone" to tone))

    object VerificationStarted : AnalyticsEvent("verification_started", emptyMap())

    class VerificationSucceeded(similarityPct: Int) :
        AnalyticsEvent("verification_succeeded", mapOf("similarity_pct" to similarityPct.toString()))

    class VerificationFailed(reason: String) :
        AnalyticsEvent("verification_failed", mapOf("reason" to reason))

    class ReportFiled(reason: String) :
        AnalyticsEvent("report_filed", mapOf("reason" to reason))

    object BlockUser : AnalyticsEvent("block_user", emptyMap())

    class SettingsToggled(propName: String, enabled: Boolean) :
        AnalyticsEvent("settings_toggled", mapOf("name" to propName, "enabled" to enabled.toString()))

    class LayerSwitched(from: String, to: String) :
        AnalyticsEvent("layer_switched", mapOf("from_layer" to from, "to_layer" to to))

    class AppOpenedFromPush(pushType: String) :
        AnalyticsEvent("app_opened_from_push", mapOf("push_type" to pushType))
}

interface AnalyticsTransport {
    fun track(name: String, properties: Map<String, String>)
}

class AnalyticsService {
    var optedIn: Boolean = false
    private var transport: AnalyticsTransport? = null

    fun attach(t: AnalyticsTransport?) { transport = t }

    fun track(event: AnalyticsEvent) {
        if (!optedIn) return
        transport?.track(event.name, event.properties)
    }

    companion object { val shared = AnalyticsService() }
}

File: apps/DriftDate/shared/DriftCore/Sources/DriftCore/Observability/AnalyticsService.swift
import Foundation

/// Where an event came from. Mirrors the Kotlin enum.
public enum Surface: String, Sendable, Codable {
    case app
    case notificationExtension = "notification_extension"
    case watch
    case complication
    case tile
}

/// Drift's canonical event taxonomy. Mirrors the Kotlin sealed class.
public enum AnalyticsEvent: Equatable, Sendable {
    case onboardingCompleted
    case waveSent(layer: Layer, surface: Surface)
    case waveMatched(layer: Layer, timeToMatchSeconds: Int)
    case chatScreenOpen(conversationId: UUID, tone: Tone)
    case replySuggestionUsed(tone: Tone, kind: ReplyKind)
    case replySuggestionDismissed(tone: Tone)
    case verificationStarted
    case verificationSucceeded(similarityPct: Int)
    case verificationFailed(reason: String)
    case reportFiled(reason: String)
    case blockUser
    case settingsToggled(name: String, enabled: Bool)
    case layerSwitched(from: Layer, to: Layer)
    case appOpenedFromPush(pushType: PushType)

    public enum ReplyKind: String, Sendable, Codable {
        case casual, context, playful
    }
    public enum PushType: String, Sendable, Codable {
        case match, message
    }

    public var name: String {
        switch self {
        case .onboardingCompleted:        return "onboarding_completed"
        case .waveSent:                   return "wave_sent"
        case .waveMatched:                return "wave_matched"
        case .chatScreenOpen:             return "chat_screen_open"
        case .replySuggestionUsed:        return "reply_suggestion_used"
        case .replySuggestionDismissed:   return "reply_suggestion_dismissed"
        case .verificationStarted:        return "verification_started"
        case .verificationSucceeded:      return "verification_succeeded"
        case .verificationFailed:         return "verification_failed"
        case .reportFiled:                return "report_filed"
        case .blockUser:                  return "block_user"
        case .settingsToggled:            return "settings_toggled"
        case .layerSwitched:              return "layer_switched"
        case .appOpenedFromPush:          return "app_opened_from_push"
        }
    }

    public var properties: [String: String] {
        switch self {
        case .onboardingCompleted, .verificationStarted, .blockUser:
            return [:]
        case .waveSent(let layer, let surface):
            return ["layer": layer.rawValue, "surface": surface.rawValue]
        case .waveMatched(let layer, let secs):
            return ["layer": layer.rawValue, "time_to_match_seconds": String(secs)]
        case .chatScreenOpen(let id, let tone):
            return ["conversation_id": id.uuidString, "tone": tone.rawValue]
        case .replySuggestionUsed(let tone, let kind):
            return ["tone": tone.rawValue, "kind": kind.rawValue]
        case .replySuggestionDismissed(let tone):
            return ["tone": tone.rawValue]
        case .verificationSucceeded(let pct):
            return ["similarity_pct": String(pct)]
        case .verificationFailed(let reason):
            return ["reason": reason]
        case .reportFiled(let reason):
            return ["reason": reason]
        case .settingsToggled(let name, let enabled):
            return ["name": name, "enabled": enabled ? "true" : "false"]
        case .layerSwitched(let from, let to):
            return ["from_layer": from.rawValue, "to_layer": to.rawValue]
        case .appOpenedFromPush(let pushType):
            return ["push_type": pushType.rawValue]
        }
    }
}

public protocol AnalyticsTransport: AnyObject {
    func track(name: String, properties: [String: String])
}

public final class AnalyticsService {
    public static let shared = AnalyticsService()

    public var optedIn: Bool = false
    private weak var transport: AnalyticsTransport?

    public func attach(_ transport: AnalyticsTransport?) { self.transport = transport }

    public func track(_ event: AnalyticsEvent) {
        guard optedIn, let transport else { return }
        transport.track(name: event.name, properties: event.properties)
    }

    #if canImport(PostHog)
    /// Wires up PostHog. Called from the host app's AppDelegate when
    /// `POSTHOG_API_KEY` is non-empty.
    public func usePostHog(apiKey: String, host: String) {
        // The app target is responsible for creating the actual PostHog
        // configuration. This stub keeps the API surface stable on
        // open-source builds with no SDK installed.
    }
    #endif
}
